tests/proptest.rs

```diff
@@ -2,35 +2,39 @@
 
 use std::{collections::BTreeSet as Set, error::Error};
 
+use pubgrub::error::PubGrubError;
+use pubgrub::package::Package;
+use pubgrub::range::Range;
+use pubgrub::report::{DefaultStringReporter, Reporter};
+use pubgrub::solver::{
+    choose_package_with_fewest_versions, resolve, Dependencies, DependencyProvider,
+    OfflineDependencyProvider,
+};
+use pubgrub::version::{NumberVersion, Version};
+
 use proptest::collection::{btree_map, vec};
 use proptest::prelude::*;
 use proptest::sample::Index;
 use proptest::string::string_regex;
 
-use pubgrub::range::Range;
-use pubgrub::solver::{resolve, Dependencies, DependencyProvider, OfflineDependencyProvider};
-use pubgrub::{
-    error::PubGrubError, package::Package, report::DefaultStringReporter, report::Reporter,
-    version::NumberVersion, version::Version,
-};
-
 use crate::sat_dependency_provider::SatResolve;
 
 mod sat_dependency_provider;
 
-/// The same as DP but takes versions from the opposite end:
-/// if DP returns versions from newest to oldest, this returns them from oldest to newest.
+/// The same as [OfflineDependencyProvider] but takes versions from the opposite end:
+/// if [OfflineDependencyProvider] returns versions from newest to oldest, this returns them from oldest to newest.
 #[derive(Clone)]
-struct ReverseDependencyProvider<DP>(DP);
-
-impl<P: Package, V: Version, DP: DependencyProvider<P, V>> DependencyProvider<P, V>
-    for ReverseDependencyProvider<DP>
-{
-    fn list_available_versions(&self, p: &P) -> Result<Vec<V>, Box<dyn Error>> {
-        self.0.list_available_versions(p).map(|mut v| {
-            v.reverse();
-            v
-        })
+struct OldestVersionsDependencyProvider<P: Package, V: Version>(OfflineDependencyProvider<P, V>);
+
+impl<P: Package, V: Version> DependencyProvider<P, V> for OldestVersionsDependencyProvider<P, V> {
+    fn make_decision<T: std::borrow::Borrow<P>, U: std::borrow::Borrow<Range<V>>>(
+        &self,
+        potential_packages: impl Iterator<Item = (T, U)>,
+    ) -> Result<(T, Option<V>), Box<dyn Error>> {
```

Comment:
As we note in the docs:
>   /// The strategy employed to choose such package and version
    /// cannot change the existence of a solution or not,

So why should it be possible to return an error here?

What we can do is to return `Option<(T, Option<V>)>` here instead of the result.
Then change `choose_package_with_fewest_versions` to return an `Option` as well and instead of
```
        .expect("potential_packages gave us an empty iterator");
```
we can just return `None`.

Alternatively we get don't introduce the `Option` in this type signature (keeping just `(T, Option<V>)`) and we can leave the above mentioned panic in.

Thoughts?

Reply:
There are lots of things that can go wrong while giving control back to the implementer of `make_decision` like when reading files on disk. Without a result, we force the implementer to panic inside `make_decision`. With a result, they can handle the error that is propagated to `resolved` however they want.

Reply:
Makes sense.